source/server/admin/admin_filter.cc

```diff
@@ -68,14 +68,18 @@ void AdminFilter::onComplete() {
   Buffer::OwnedImpl response;
   auto header_map = Http::ResponseHeaderMapImpl::create();
   RELEASE_ASSERT(request_headers_, "");
-  Http::Code code = admin_server_callback_func_(path, *header_map, response, *this);
-  Utility::populateFallbackResponseHeaders(code, *header_map);
-  decoder_callbacks_->encodeHeaders(std::move(header_map),
-                                    end_stream_on_complete_ && response.length() == 0,
-                                    StreamInfo::ResponseCodeDetails::get().AdminFilterResponse);
-
-  if (response.length() > 0) {
-    decoder_callbacks_->encodeData(response, end_stream_on_complete_);
+  for (bool cont = true, first = true; cont; first = false) {
```

Comment:
This pattern here is deeply flawed on a couple of levels. This loops over all the data, streaming it out to the client. Consider the data could be 1G. That will block the main thread until all the data gets to the client.

It also will stuff this into the network without any notion of client flow-control.

Instead we should use the exiting Envoy filter state-machine stuff to stream out a chunk at a time and return control.

See https://github.com/envoyproxy/envoy/blob/3d5f63705553d255e37777e4f4d0f69185c29654/source/extensions/filters/http/decompressor/decompressor_filter.cc#L96 (gzip decompress filter) for something that I *think* is doing this right.

OTOH I tested this manually and it seems to "work" in the sense that all the data comes out and is received by 'wget' via chunked encoding.

Also what we really want here is create a Handler from the callback and save it in the filter, and then call the handler for each chunk whenever encodeData is called (I think). I am not too fluent in Envoy filter streaming anymore.